nnsvs/mdn.py

```diff
@@ -0,0 +1,169 @@
+
+    Implementation references:
+    1. Mixture Density Networks by Mike Dusenberry https://mikedusenberry.com/mixture-density-networks
+    2. PRML book https://www.microsoft.com/en-us/research/people/cmbishop/prml-book/
+    3. sagelywizard/pytorch-mdn https://github.com/sagelywizard/pytorch-mdn
+    4. sksq96/pytorch-mdn https://github.com/sksq96/pytorch-mdn
+
+    Arguments:
+        in_dim (int): the number of dimensions in the input
+        out_dim (int): the number of dimensions in the output
+        num_gaussians (int): the number of mixture component
+    Input:
+        minibatch (B, T, D_in): B is the batch size and T is data lengths of this batch, 
+            and D_in is in_dim.
+    Output:
+        log_pi, log_sigma, mu (B, T, G), (B, T, G, D_out), (B, T, G, D_out): 
+            G is num_gaussians and D_out is out_dim.
+            log_pi is log of a multinomial distribution of the Gaussians. 
+            mu and log_sigma are the mean and the log of standard deviation of each Gaussian.
+    """
+    def __init__(self, in_dim, out_dim, num_gaussians=30):
+        super(MDNLayer, self).__init__()
+        self.in_dim = in_dim
+        self.out_dim = out_dim
+        self.num_gaussians = num_gaussians
+
+        self.log_pi = nn.Linear(in_dim, num_gaussians)
+        self.log_sigma = nn.Linear(in_dim, out_dim * num_gaussians)
+        self.mu = nn.Linear(in_dim, out_dim * num_gaussians)
+
+    def forward(self, minibatch):
+        log_pi = F.log_softmax(self.log_pi(minibatch), dim=2)
+        log_sigma = self.log_sigma(minibatch)
+        log_sigma = log_sigma.view(len(minibatch), -1, self.num_gaussians, self.out_dim)        
+        mu = self.mu(minibatch)
+        mu = mu.view(len(minibatch), -1, self.num_gaussians, self.out_dim)
+        return log_pi, log_sigma, mu
+
+def mdn_loss(log_pi, log_sigma, mu, target, log_clamp_min=-7.0, reduce=True):
+    """Calculates the error, given the MoG parameters and the target.
+    The loss is the negative log likelihood of the data given the MoG
+    parameters.
+    Arguments:
+        log_pi (B, T, G): The log of multinomial distribution of the Gaussians. B is the batch size,
+            T is data length of this batch, and G is num_gaussians of class MDNLayer.
+        log_sigma (B, T , G ,D_out): The log standard deviation of the Gaussians. D_out is out_dim of class 
+            MDNLayer.
+        mu (B , T, G, D_out): The means of the Gaussians. 
+        target (B, T, D_out): The target variables.
+        log_clamp_min: minimum value of logged values(log_pi, log_sigma, and torch.distributions.Normal.log_prob)
+        reduce: If True, the losses are averaged for each batch.
+    Returns:
+        loss (B) or (B, T): Negative Log Likelihood of Mixture Density Networks.
+    """
+
+    # Clip log_sigma and log_pi with log_clamp_min for numerical stability
+    log_sigma = torch.clamp(log_sigma, min=log_clamp_min)
+    log_pi = torch.clamp(log_pi, min=log_clamp_min)
```

Comment:
I am not sure if it's right to clamp "probability" and "variance" with the same value. Would it be better to have separate thresholds?

Reply:
I think any value which can avoid producing nan is appropriate because clamping is only for numerical stability, not for any mathematical meanings.  But I split them into separate threshoulds in the latest commit for the future when we may want to specify the different values.